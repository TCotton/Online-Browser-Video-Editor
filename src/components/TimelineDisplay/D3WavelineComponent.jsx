import React, {useEffect, useRef} from 'react';
import PropTypes from 'prop-types';
import * as d3 from 'd3';

export const D3WavelineComponent = (props) => {
    const {data, width, height} = props;

    const ref = useRef();

    useEffect(() => {

        //TODO: use method as advised on the GitHub account: https://github.com/bbc/waveform-data.js/blob/master/README.md
        if (data && Array.isArray(data.min_array) && Array.isArray(data.max_array)) {

            const layout = d3.select(ref.current);

            const min = data.min_array;
            const max = data.max_array;

            const xScale = d3.scaleLinear()
                .range([0, width])
                .domain([0, data.length]);

            const yScale = d3.scaleLinear()
                .range([height, 0])
                .domain([d3.min(min), d3.max(max)]);

            const d3Line = d3.line()
                .x( (d, i) => {
                    return xScale(i);
                })
                .y( (d) => {
                    return yScale(d);
                });

            layout
                .datum(max)
                .attr("d", d3Line);
        }

    }, [data])

    return (
        <svg
            viewBox={`0 0 ${width} ${height}`}
            preserveAspectRatio="xMidYMid slice"
            data-testid="svg"
        >
            <g className="area" transform="translate(0,0)" data-testid="area">
                <path
                    className="wave"
                    ref={ref}
                    fill="#20e0bb"
                />
            </g>
        </svg>
    )

}

D3WavelineComponent.propTypes = {
    height: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    data: PropTypes.shape({
        min_array: PropTypes.arrayOf(PropTypes.number).isRequired,
        max_array: PropTypes.arrayOf(PropTypes.number).isRequired,
        length: PropTypes.number.isRequired,
    }).isRequired
};