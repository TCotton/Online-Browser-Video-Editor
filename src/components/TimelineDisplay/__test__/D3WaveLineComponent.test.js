import React from 'react'
import { render } from '@testing-library/react'
import { D3WavelineComponent } from '../D3WavelineComponent'

describe('D3WavelineComponent', () => {
  const props = {
    height: 999,
    width: 999,
    data: {
      min_array: [1],
      max_array: [2],
      length: 123
    }
  }

  it('should render successfully', () => {
    const { container } = render(<D3WavelineComponent {...props} />)
    expect(container.querySelector('[data-testid="svg"]')).not.toBeNull()
    expect(container.querySelector('[data-testid="area"]')).not.toBeNull()
  })

  it('should match snapshot', () => {
    const { baseElement } = render(<D3WavelineComponent {...props} />)
    expect(baseElement).toMatchSnapshot()
  })
})
