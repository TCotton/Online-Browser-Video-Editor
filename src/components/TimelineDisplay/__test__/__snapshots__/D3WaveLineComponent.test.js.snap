// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`D3WavelineComponent should match snapshot 1`] = `
<body>
  <div>
    <svg
      data-testid="svg"
      height="999"
      preserveAspectRatio="xMidYMid slice"
      viewBox="0 0 999 999"
      width="999"
    >
      <g
        class="area"
        data-testid="area"
        transform="translate(0,0)"
      >
        <path
          class="wave"
          d="M0,0L0,999Z"
          transform="translate(0,0)"
        />
      </g>
    </svg>
  </div>
</body>
`;
